### PR TITLE
Add cross-workspace 4-part naming for FabricSpark

### PR DESCRIPTION
**Pull request by sdebruyn**
## Summary

Closes #233.

- Adds a `workspace` field to `FabricSparkRelation` that, when set, prepends a backtick-quoted workspace name to produce 4-part names (`workspace.lakehouse.schema.table`) in rendered SQL
- `create_from()` pulls `workspace_name` from model config automatically
- `incorporate()` preserves the workspace field across relation copies
- Workspace prefix is only emitted when `include_policy.database` is True — CTEs and temp tables that use `.include(database=False)` remain unqualified
- Snapshot materialization propagates workspace to both target and staging relations
- The field is render-only and does not participate in cache identity

## Test plan

- [x] Unit tests for 4-part rendering (with workspace, without, with database excluded)
- [x] Unit tests for `incorporate()` preserving, setting, and overriding workspace
- [x] Unit tests for `create_from()` pulling `workspace_name` from config and kwarg precedence
- [x] All 449 existing unit tests pass without regressions
- [ ] Integration test with cross-workspace snapshot (requires real Fabric infrastructure)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/dbt/adapters/fabricspark/fabricspark_relation.py [hunk 1]
```diff
@@ -63,6 +65,24 @@ class FabricSparkRelation(BaseRelation):
...
+        relation_config: RelationConfig,
+        **kwargs: Any,
+    ) -> "FabricSparkRelation":
+        if "workspace" not in kwargs:
+            cfg = getattr(relation_config, "config", None)
+            if cfg is not None:
+                try:
+                    ws_name = cfg.get("workspace_name")
+                except Exception:
+                    ws_name = None
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_relation.py:83]: Fixed — narrowed the except to (AttributeError, TypeError) in a0fcacf.

File: src/dbt/adapters/fabricspark/fabricspark_relation.py [hunk 2]
```diff
@@ -79,6 +99,18 @@ def try_translate_type(cls, relation_type: str | None) -> FabricSparkRelationTyp
     def get_relation_type(cls) -> builtins.type[FabricSparkRelationType]:
         return FabricSparkRelationType
 
+    def render(self) -> str:
+        base = super().render()
+        if self.workspace and self.include_policy.get_part(ComponentName.Database):
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_relation.py]: Fixed in a0fcacf — `render()` now also checks `self.database` before prepending workspace. If database is `None` (e.g. after `without_identifier()` or similar), the workspace is omitted since a 4-part name without the database component is invalid.

File: src/dbt/adapters/fabricspark/fabricspark_relation.py [hunk 3]
```diff
@@ -79,6 +99,18 @@ def try_translate_type(cls, relation_type: str | None) -> FabricSparkRelationTyp
...
+            quoted_ws = self.quoted(self.workspace)
+            return f"{quoted_ws}.{base}" if base else quoted_ws
+        return base
+
+    def incorporate(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_relation.py]: Fixed in a0fcacf — added `**kwargs: Any` type hint and `-> "FabricSparkRelation"` return type. Also added a test (`test_incorporate_can_clear_workspace`) that verifies `incorporate(workspace=None)` explicitly clears the workspace from a relation that previously had one.

File: src/dbt/adapters/fabricspark/fabricspark_relation.py [hunk 4]
```diff
@@ -46,6 +47,7 @@ class FabricSparkRelation(BaseRelation):
     quote_character: str = "`"
     require_alias: bool = False
     information: str | None = None
+    workspace: str | None = None
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_relation.py:51]: This is intentional — `catalog` and `workspace` serve different purposes:

- `catalog` in `_build_spark_relation_list` is the Spark catalog name from `SHOW CATALOGS`. It's used as the `database` component for cache identity and metadata queries. This is how dbt-spark works — the catalog maps to `relation.database`.
- `workspace` is a rendering-only concept for cross-workspace SQL. It's set per-model via `config(workspace_name=...)` in `create_from()`, not from catalog metadata.

The flow is: metadata queries use `catalog` → stored in cache as `relation.database` → when rendering SQL, if `workspace` is set, it prepends the workspace to make a 4-part name. They're orthogonal. A relation from the current workspace has `database=lakehouse_name` (from catalog) and `workspace=None`. A cross-workspace relation has `database=lakehouse_name` and `workspace=other_workspace`.

File: src/dbt/include/fabricspark/macros/materializations/snapshots/snapshot.sql [hunk 5]
```diff
@@ -8,13 +8,20 @@
...
   {% set target_relation_exists, target_relation = get_or_create_relation(
           database=model.database,
           schema=model.schema,
           identifier=target_table,
           type='table') -%}
 
+  {# Propagate workspace for cross-workspace 4-part naming #}
+  {%- if workspace_name -%}
+    {% set target_relation = target_relation.incorporate(workspace=workspace_name) %}
+  {%- endif -%}
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/include/fabricspark/macros/materializations/snapshots/snapshot.sql:23]: This is necessary because `get_or_create_relation` (from dbt-core's base snapshot materialization) calls `adapter.Relation.create()` directly, not `create_from()`. Since `create()` doesn't go through our `create_from()` override, the returned relation has no `workspace` — it needs to be re-incorporated from the config.

This is the same approach taken in microsoft/dbt-fabricspark PR #182 (the upstream implementation of this feature). The alternative would be to override `get_or_create_relation` in the adapter class, but that's more invasive and fragile.

File: src/dbt/include/fabricspark/macros/materializations/snapshots/snapshot.sql [hunk 6]
```diff
@@ -8,13 +8,20 @@
...
   {# dbt-spark: database=none -- FabricSpark supports 3-part names #}
   {% set target_relation_exists, target_relation = get_or_create_relation(
           database=model.database,
           schema=model.schema,
           identifier=target_table,
           type='table') -%}
 
+  {# Propagate workspace for cross-workspace 4-part naming #}
+  {%- if workspace_name -%}
+    {% set target_relation = target_relation.incorporate(workspace=workspace_name) %}
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/include/fabricspark/macros/materializations/snapshots/snapshot.sql:22]: Acknowledged — \`target_relation_exists\` is evaluated before the workspace is incorporated onto \`target_relation\`. However, this is correct behavior: the existence check queries the _current_ workspace's metadata (via \`adapter.get_relation\`), which doesn't know about cross-workspace relations. The workspace incorporation happens after because it only affects SQL rendering, not metadata lookup.

The upstream PR #182 has the same ordering for the same reason: existence checks are local, workspace is a rendering concern. A true cross-workspace existence check would require executing SQL against the remote workspace, which the Livy session may not have permissions for.

This is a known limitation — if the snapshot target is in another workspace and already exists, we won't detect it during the existence check. This matches upstream behavior.

File: src/dbt/include/fabricspark/macros/materializations/models/table.sql [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+  {%- call statement('main', language=language) -%}
+    {{ create_table_as(False, target_relation, compiled_code, language) }}
+  {%- endcall -%}
+
+  {% set should_revoke = should_revoke(old_relation, full_refresh_mode=True) %}
+  {% do apply_grants(target_relation, grant_config, should_revoke) %}
+
+  {% do persist_docs(target_relation, model) %}
+
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/include/fabricspark/macros/materializations/models/table.sql:26]: This matches dbt-spark's table materialization, which also calls `run_hooks(pre_hooks)` / `run_hooks(post_hooks)` once without the `inside_transaction` split (see dbt-spark table.sql lines 12 and 44). The split is present in snapshot.sql because that one is based on dbt-core's snapshot materialization which has transaction semantics. Spark has no transactions, so `inside_transaction` is a no-op — but we keep it in snapshot.sql for consistency with the base we forked from.

File: src/dbt/include/fabricspark/macros/materializations/models/table.sql [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  {% do persist_constraints(target_relation, model) %}
+
+  {{ run_hooks(post_hooks) }}
+
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/include/fabricspark/macros/materializations/models/table.sql:30]: This file is necessary for cross-workspace 4-part naming to work. dbt-spark's table materialization uses `api.Relation.create(identifier, schema, database, type='table')` which creates a fresh relation without the `workspace` field — making it impossible for the rendered SQL to include the workspace prefix. The fix is `this.incorporate(type='table')` which preserves workspace from the model config.

The two deviations from dbt-spark are documented with inline comments:
1. `this.incorporate()` instead of `api.Relation.create()` — preserves workspace
2. Simplified drop logic instead of delta/iceberg detection — Fabric Lakehouse is always Delta and doesn't support `CREATE OR REPLACE TABLE` (only `CREATE OR REPLACE MATERIALIZED LAKE VIEW`)

Features like `partition_by` and `file_format` are handled in `create_table_as`, not in the materialization itself, so they're unaffected.

File: src/dbt/include/fabricspark/macros/materializations/snapshots/snapshot.sql [hunk 9]
```diff
@@ -8,18 +8,25 @@
...
+  {%- endif -%}
+
   {%- if not target_relation.is_table -%}
     {% do exceptions.relation_wrong_type(target_relation, 'table') %}
   {%- endif -%}
 
-  {% if not adapter.check_schema_exists(model.database, model.schema) %}
+  {% if not workspace_name and not adapter.check_schema_exists(model.database, model.schema) %}
     {% do create_schema(target_relation) %}
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/include/fabricspark/macros/materializations/snapshots/snapshot.sql:33]: Added an explanatory comment in 6eb007d. The root cause is that `create_schema` calls `relation.without_identifier()`, and `render()` intentionally requires `self.identifier` to prepend workspace (otherwise local metadata operations like `check_schema_exists` would incorrectly target the remote workspace).

Cross-workspace schema creation works via raw SQL (as demonstrated in the integration tests' `create_cross_workspace_schema` macro). Users must ensure the target schema exists before running cross-workspace models — this matches the upstream PR #182 approach and is documented in the snapshot macro comment.

File: src/dbt/adapters/fabricspark/fabricspark_relation.py [hunk 10]
```diff
@@ -83,6 +103,24 @@ def try_translate_type(cls, relation_type: str | None) -> FabricSparkRelationTyp
...
+        if (
+            self.workspace
+            and self.identifier
+            and self.database
+            and self.include_policy.get_part(ComponentName.Database)
+            and self.include_policy.get_part(ComponentName.Identifier)
+        ):
+            quoted_ws = self.quoted(self.workspace)
+            return f"{quoted_ws}.{base}" if base else quoted_ws
+        return base
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply src/dbt/adapters/fabricspark/fabricspark_relation.py:119]: This is intentional — added a code comment in 6eb007d explaining the design.

The identifier gate exists because `without_identifier()` is used by `check_schema_exists`, `create_schema`, and other metadata operations that must target the *current* workspace (the one the Livy session is bound to), even when a model has `workspace_name` configured. Without this gate, those operations would generate e.g. `CREATE SCHEMA \`remote_ws\`.\`lh\`.\`schema\`` when they should only operate locally.

Cross-workspace schema creation is handled separately via raw SQL in user macros or on-run-start hooks. The `self.database` check alone isn't sufficient because `without_identifier()` preserves database — it's specifically the absence of identifier that signals a schema-level operation.

File: tests/fabricspark/adapter/test_cross_workspace.py [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+        return {
+            "vars": {
+                "cross_workspace_name": cross_workspace_config["workspace_name"],
+                "cross_lakehouse_name": cross_workspace_config["lakehouse_name"],
+            },
+            "on-run-start": [
+                "{{ drop_cross_workspace_objects() }}",
+                "{{ create_cross_workspace_source() }}",
+            ],
+        }
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply tests/fabricspark/adapter/test_cross_workspace.py:377]: Fixed in 6eb007d — the snapshot test's `on-run-start` now calls a scoped `drop_snapshot_object()` macro that only drops `cross_ws_snapshot` (the one object this class creates), instead of the shared `drop_cross_workspace_objects()` which drops objects from all test classes.

File: tests/conftest.py [hunk 12]
```diff
@@ -276,6 +285,14 @@ def purview_client(
     return PurviewClient(credentials.purview_endpoint, fabric_token_provider)
 
 
+@pytest.fixture(scope="class")
+def cross_workspace_config():
+    return {
+        "workspace_name": os.getenv("FABRIC_TEST_CROSS_WORKSPACE_NAME"),
+        "lakehouse_name": os.getenv("FABRIC_TEST_CROSS_LAKEHOUSE_NAME"),
```
[the remark this answers is not in the document]
Comment (sdebruyn) [reply tests/conftest.py:296]: Fixed in 6eb007d — `has_cross_workspace` now requires both `FABRIC_TEST_CROSS_WORKSPACE_NAME` and `FABRIC_TEST_CROSS_LAKEHOUSE_NAME` to be set. Tests are skipped if either is missing.
